### ISSUE TITLE
error: array lengths can't depend on generic parameters

### ISSUE THREAD

**Issue by jonathanstrong** [bluss/matrixmultiply#50]
ran into this error with nightly (via `$ cargo check` on master at 4ac62c28):

```console
error: array lengths can't depend on generic parameters
   --> src/sgemm_kernel.rs:223:40
    |
223 |     let mut ab = [_mm256_setzero_ps(); MR];
    |                                        ^^
```

```console
$ rustc --version
rustc 1.42.0-nightly (a9dd56ff9 2019-12-30)
```

haven't been able to make much headway in understanding what changed in rustc, but it seems likely this is a compiler bug. 

It looks like substituting a `const fn` implemented on the struct does work (current implementation seems to rely on trait associated constants). 

**update**: for a quick workaround, roll-back to nightly-2019-12-25 (ho ho ho!)

**Comment by azriel91** [comment]
Related issue: https://github.com/rust-lang/rust/issues/67743

**Comment by norru** [comment]
Reproduced in:
- rustc 1.42.0-nightly (859764425 2020-01-07)

Works in:
- rustc 1.40.0 (73528e339 2019-12-16)

**Comment by bluss** [comment]
Thanks! As you know, this code compiles since Rust 1.28, so it's unfortunate that it is broken on nightly.

**Comment by mvlabat** [comment]
I was able to compile it on rustc 1.42.0-nightly (6d3f4e0aa 2020-01-25), Windows 10

**Comment by bluss** [comment]
The issue has been fixed in Rust and is fixed in Rust 1.42. It will continue to be an issue (unless further changes are done in matrixmultiply) with older versions, but it is probably too late to matter now.
